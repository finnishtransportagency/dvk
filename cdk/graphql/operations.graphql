query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
