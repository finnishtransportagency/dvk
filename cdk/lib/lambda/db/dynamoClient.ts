import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';

let client: DynamoDBDocumentClient;

function getDynamoDBDocumentClient(): DynamoDBDocumentClient {
  if (!client) {
    client = DynamoDBDocumentClient.from(new DynamoDBClient({ region: 'eu-west-1' }), { marshallOptions: { removeUndefinedValues: true } });
  }
  return client;
}

export { getDynamoDBDocumentClient };
