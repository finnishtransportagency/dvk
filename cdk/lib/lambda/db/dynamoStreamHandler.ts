import { DynamoDBStreamEvent, DynamoDBRecord, DynamoDBStreamHandler } from 'aws-lambda/trigger/dynamodb-stream';
import { AttributeValue } from '@aws-sdk/client-dynamodb';
import { unmarshall } from '@aws-sdk/util-dynamodb';
import { auditLog, log } from '../logger';
import { DeleteObjectCommand, PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import FairwayCardDBModel from './fairwayCardDBModel';
import { getNewStaticBucketName } from '../environment';

const s3Client = new S3Client({ region: 'eu-west-1' });

// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function saveToS3(data: Record<string, any>, bucket: string) {
  const command = new PutObjectCommand({
    Key: data.id + '.json',
    Bucket: bucket,
    Body: JSON.stringify(data),
    ContentType: 'application/json',
  });

  log.info(`saving ${data.id}.json to s3://${bucket}`);
  log.debug(command);

  await s3Client.send(command);
}

function getBucketForTable(arn: string | undefined) {
  if (!arn) {
    return undefined;
  } else if (arn.includes('FairwayCard')) {
    return { bucket: process.env.FAIRWAYCARD_VERSIONING_BUCKET, table: 'FairwayCard' };
  } else if (arn.includes('Harbor')) {
    return { bucket: process.env.HARBOR_VERSIONING_BUCKET, table: 'Harbor' };
  }
  return undefined;
}

async function handleUpdate(record: DynamoDBRecord) {
  log.info('handling event from %s', record.eventSourceARN);

  if (!record.dynamodb?.NewImage) {
    log.error('no new record to save to s3');
    return;
  }

  const bucket = getBucketForTable(record.eventSourceARN);
  if (!bucket?.bucket) {
    log.error('no known table found %s', record.eventSourceARN);
    return;
  }

  const data = unmarshall(record.dynamodb.NewImage as { [key: string]: AttributeValue });

  await saveToS3(data, bucket.bucket);
}

async function handleRemove(record: DynamoDBRecord) {
  log.info('handling remove event from %s', record.eventSourceARN);
  if (!record.dynamodb?.OldImage) {
    log.error('no old record');
    return;
  }
  const bucket = getBucketForTable(record.eventSourceARN);
  if (!bucket?.bucket) {
    log.error('no known table found %s', record.eventSourceARN);
    return;
  }

  const data = unmarshall(record.dynamodb.OldImage as { [key: string]: AttributeValue }) as FairwayCardDBModel;
  for (const picture of data.pictures ?? []) {
    const command = new DeleteObjectCommand({
      Key: data.id + '/' + picture.id,
      Bucket: getNewStaticBucketName(),
    });
    try {
      log.debug('removing picture %s', picture.id);
      await s3Client.send(command);
    } catch (e) {
      log.error('removing picture %s failed: %s', picture.id, e);
    }
  }

  auditLog.info({ data, user: 'system' }, `${bucket.table} removed`);
}

const handler: DynamoDBStreamHandler = async function (event: DynamoDBStreamEvent) {
  event.Records.forEach((record) => {
    log.debug('Event source: %s', record.eventSource);
    log.debug('Event sourceARN: %s', record.eventSourceARN);
    log.debug('Event Id: %s', record.eventID);
    log.debug('Event Type: %s', record.eventName);
    log.debug('DynamoDB Record: %j', record.dynamodb);
  });

  for (const record of event.Records) {
    switch (record.eventName) {
      case 'INSERT':
      case 'MODIFY':
        await handleUpdate(record);
        break;
      case 'REMOVE':
        await handleRemove(record);
    }
  }
};

export { handler };
