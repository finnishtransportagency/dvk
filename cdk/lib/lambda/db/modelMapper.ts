import { FeatureCollection } from 'geojson';
import {
  Fairway,
  FairwayCard,
  GeometryInput,
  Harbor,
  InputMaybe,
  Maybe,
  OperationError,
  PilotPlace,
  TextInput,
  TrafficService,
} from '../../../graphql/generated';
import { CurrentUser } from '../api/login';
import { fetchPilotPoints } from '../api/traficom';
import { getFromCache, cacheResponse, CacheResponse } from '../graphql/cache';
import { log } from '../logger';
import FairwayCardDBModel, { FairwayDBModel, PilotRoute, TemporaryNotification, TrafficServiceDBModel } from './fairwayCardDBModel';
import HarborDBModel from './harborDBModel';
import { fetchPilotRouteData } from '../api/pilotRoutes';
import { saveResponseToS3 } from '../util';
import zlib from 'zlib';

const MAX_TEXT_LENGTH = 2000;
const MAX_NUMBER_LENGTH = 10;
// geometry allows only 5 decimals
const MAX_GEOMETRY_NUMBER_LENGTH = 8;

function checkLength(maxLength: number, ...text: string[]) {
  text.forEach((s) => {
    if (s && s.length > maxLength) {
      throw new Error(OperationError.InvalidInput);
    }
  });
}

export function mapNumber(text: Maybe<string> | undefined, maxLength = MAX_NUMBER_LENGTH): number | null {
  if (text) {
    checkLength(maxLength, text);
    const number = Number(text.replace(',', '.'));
    if (Number.isNaN(number)) {
      throw new Error(OperationError.InvalidInput);
    }
    if (text.trim().length > 0) {
      return number;
    }
  }
  return null;
}

export function mapGeometry(geometry?: InputMaybe<GeometryInput>, maxLength = MAX_GEOMETRY_NUMBER_LENGTH) {
  if (geometry?.lat && geometry.lon) {
    const geom = { type: 'Point', coordinates: [mapNumber(geometry.lon, maxLength), mapNumber(geometry.lat, maxLength)] };
    if (geom.coordinates[0] && geom.coordinates[1]) {
      if (geom.coordinates[0] >= 32 || geom.coordinates[0] < 17 || geom.coordinates[1] >= 70 || geom.coordinates[1] < 58) {
        throw new Error(OperationError.InvalidInput);
      }
      return geom;
    }
    throw new Error(OperationError.InvalidInput);
  }
  return null;
}

export function mapId(text?: Maybe<string>) {
  if (text && text.trim().length > 0) {
    checkLength(200, text);
    const m = /^[a-z]+[a-z\d]*$/.exec(text);
    if (m && m[0].length === text.length) {
      return text;
    }
  }
  throw new Error(OperationError.InvalidInput);
}

export function mapText(text?: Maybe<TextInput>, maxLength = MAX_TEXT_LENGTH) {
  if (text?.fi && text.sv && text.en) {
    checkLength(maxLength, text.fi, text.sv, text.en);
    return {
      fi: text.fi,
      sv: text.sv,
      en: text.en,
    };
  } else if (text && (text.fi || text.sv || text.en)) {
    throw new Error(OperationError.InvalidInput);
  } else {
    return null;
  }
}

export function mapMandatoryText(text?: TextInput, maxLength = MAX_TEXT_LENGTH) {
  if (text?.fi && text?.sv && text?.en) {
    checkLength(maxLength, text.fi, text.sv, text.en);
    return {
      fi: text.fi,
      sv: text.sv,
      en: text.en,
    };
  } else {
    throw new Error(OperationError.InvalidInput);
  }
}

export function mapString(text: Maybe<string> | undefined, maxLength = MAX_TEXT_LENGTH): string | null {
  if (text) {
    checkLength(maxLength, text);
    return text;
  }
  return null;
}

export function mapInternetAddress(text: Maybe<string> | undefined): string | null {
  return mapString(text, 200);
}

function checkRegExp(regexp: RegExp, text: string) {
  if (!regexp.test(text)) {
    throw new Error(OperationError.InvalidInput);
  }
}

export function mapEmail(text: Maybe<string> | undefined): string | null {
  if (text) {
    checkRegExp(
      /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/,
      text
    );
    return text;
  }
  return null;
}

export function mapEmails(text: Maybe<Maybe<string>[]> | undefined): string[] | null {
  return text ? text.map((t) => mapEmail(t)).filter((t) => t !== null) : null;
}

export function mapPhoneNumber(text: Maybe<string> | undefined): string | null {
  if (text) {
    checkRegExp(/^[+]?\d(\s?\d){4,19}$/, text);
    return text;
  }
  return null;
}

export function mapPhoneNumbers(text: Maybe<Maybe<string>[]> | undefined): string[] | null {
  return text ? text.map((t) => mapPhoneNumber(t)).filter((t) => t !== null) : null;
}

export function mapStringArray(text: Maybe<Maybe<string>[]> | undefined, maxLength = MAX_TEXT_LENGTH): string[] | null {
  return text ? text.map((t) => mapString(t, maxLength)).filter((t) => t !== null) : null;
}

export function mapIds(ids: number[]) {
  return `#${ids.join('#')}#`;
}

export function mapFairwayIds(dbModel: FairwayCardDBModel) {
  const fairways = dbModel.fairways;
  if (fairways) {
    return mapIds(fairways.map((f) => f.id));
  }
  return '';
}

function mapNumberAndMax(text: Maybe<string> | undefined, regexp: RegExp, maxValue: number) {
  if (text) {
    checkRegExp(regexp, text);
    const number = mapNumber(text);
    if (number && (number > maxValue || number < 0)) {
      throw new Error(OperationError.InvalidInput);
    }
    return number;
  }
  return null;
}

export function mapPilotJourney(text: Maybe<string> | undefined) {
  return mapNumberAndMax(text, /^\d{1,3}[.,]?\d?$/, 999.9);
}

export function mapVhfChannel(text: Maybe<string> | undefined) {
  return mapNumberAndMax(text, /^\d{1,3}$/, 999);
}

export function mapQuayLength(text: Maybe<string> | undefined) {
  return mapNumberAndMax(text, /^\d{1,4}[.,]?\d?$/, 9999.9);
}

export function mapQuayDepth(text: Maybe<string> | undefined) {
  return mapNumberAndMax(text, /^\d{1,3}[.,]?\d{0,2}$/, 999.99);
}

export function mapVersion(text?: Maybe<string>) {
  if (text && text.trim().length > 0) {
    const regex = /^v\d+(_(latest|public))?$/;
    const passedString = regex.exec(text);
    if (passedString && passedString[0].length === text.length) {
      return text;
    }
  }
  throw new Error(OperationError.InvalidInput);
}

export const pilotPlaceMap = new Map<number, PilotPlace>();
const pilotCacheKey = 'pilotplaces';

export async function getPilotPlaceMap() {
  if (pilotPlaceMap.size === 0) {
    let response: CacheResponse | undefined;
    let data: PilotPlace[] | undefined;
    try {
      response = await getFromCache(pilotCacheKey);
      if (response.expired) {
        log.debug('fetching pilot places from api');
        data = await fetchPilotPoints();
        await cacheResponse(pilotCacheKey, data);
      } else if (response.data) {
        log.debug('parsing pilot places from cache');
        data = JSON.parse(response.data) as PilotPlace[];
      }
    } catch (e) {
      if (!data && response?.data) {
        log.warn('parsing expired pilot places from cache');
        data = JSON.parse(response.data) as PilotPlace[];
      }
    }
    data?.forEach((p) => pilotPlaceMap.set(p.id, p));
  }
  return pilotPlaceMap;
}

function mapFairwayDBModelToFairway(dbModel: FairwayDBModel): Fairway {
  const fairway: Fairway = {
    id: dbModel.id,
    primary: dbModel.primary ?? false,
    primarySequenceNumber: dbModel.primarySequenceNumber,
    secondary: dbModel.secondary ?? false,
    secondarySequenceNumber: dbModel.secondarySequenceNumber,
  };
  return fairway;
}

function mapTrafficService(service: TrafficServiceDBModel | undefined | null, pilotMap: Map<number, PilotPlace>): TrafficService {
  return {
    pilot: {
      email: service?.pilot?.email,
      extraInfo: service?.pilot?.extraInfo,
      fax: service?.pilot?.fax,
      internet: service?.pilot?.internet,
      phoneNumber: service?.pilot?.phoneNumber,
      places:
        service?.pilot?.places?.map((p) => {
          return {
            id: p.id,
            pilotJourney: p.pilotJourney,
            name: pilotMap.get(p.id)?.name ?? { fi: '', sv: '', en: '' },
            geometry: pilotMap.get(p.id)?.geometry ?? { type: 'Point', coordinates: [] },
          };
        }) ?? [],
    },
    tugs: service?.tugs ?? null,
    vts: service?.vts ?? null,
  };
}

export function mapFairwayCardDBModelToGraphqlType(
  dbModel: FairwayCardDBModel,
  pilotMap: Map<number, PilotPlace>,
  user: CurrentUser | undefined,
  pilotRoutes: FeatureCollection
) {
  const card: FairwayCard = {
    id: dbModel.id,
    version: dbModel.version,
    currentPublic: dbModel.currentPublic,
    name: {
      fi: dbModel.name?.fi,
      sv: dbModel.name?.sv,
      en: dbModel.name?.en,
    },
    n2000HeightSystem: !!dbModel.n2000HeightSystem,
    status: dbModel.status,
    group: dbModel.group,
    creator: user ? dbModel.creator : null,
    creationTimestamp: dbModel.creationTimestamp,
    modifier: user ? dbModel.modifier : null,
    modificationTimestamp: dbModel.modificationTimestamp,
    fairways: [],
    generalInfo: dbModel.generalInfo,
    anchorage: dbModel.anchorage,
    iceCondition: dbModel.iceCondition,
    attention: dbModel.attention,
    additionalInfo: dbModel.additionalInfo,
    lineText: dbModel.lineText,
    designSpeed: dbModel.designSpeed,
    navigationCondition: dbModel.navigationCondition,
    windRecommendation: dbModel.windRecommendation,
    visibility: dbModel.visibility,
    mareographs: dbModel.mareographs,
    speedLimit: dbModel.speedLimit,
    vesselRecommendation: dbModel.vesselRecommendation,
    trafficService: mapTrafficService(dbModel.trafficService, pilotMap),
    harbors: dbModel.harbors,
    pilotRoutes: mapPilotRoutes(dbModel.pilotRoutes ?? [], pilotRoutes),
    fairwayIds: mapFairwayIds(dbModel),
    pictures: dbModel.pictures,
    temporaryNotifications: mapTemporaryNotifications(dbModel.temporaryNotifications ?? []),
    latest: dbModel.latest,
    latestVersionUsed: dbModel.latestVersionUsed,
    publishDetails: dbModel.publishDetails,
  };

  if (card.fairways) {
    for (const fairway of dbModel.fairways || []) {
      card.fairways.push(mapFairwayDBModelToFairway(fairway));
    }
  }
  return card;
}

export function mapHarborDBModelToGraphqlType(dbModel: HarborDBModel, user: CurrentUser | undefined): Harbor {
  return {
    id: dbModel.id,
    version: dbModel.version,
    cargo: dbModel.cargo,
    company: dbModel.company,
    creator: user ? dbModel.creator : null,
    creationTimestamp: dbModel.creationTimestamp,
    email: dbModel.email,
    extraInfo: dbModel.extraInfo,
    fax: dbModel.fax,
    geometry: dbModel.geometry,
    harborBasin: dbModel.harborBasin,
    internet: dbModel.internet,
    modifier: user ? dbModel.modifier : null,
    modificationTimestamp: dbModel.modificationTimestamp,
    n2000HeightSystem: dbModel.n2000HeightSystem,
    name: dbModel.name,
    phoneNumber: dbModel.phoneNumber,
    quays: dbModel.quays,
    status: dbModel.status,
    latest: dbModel.latest,
    latestVersionUsed: dbModel.latestVersionUsed,
    publishDetails: dbModel.publishDetails,
  };
}

const pilotRoutesCacheKey = 'pilotroutes';

export async function getPilotRoutes() {
  // get data (from cache or api)
  let response: CacheResponse | undefined;
  let data: FeatureCollection | undefined;
  try {
    response = await getFromCache(pilotRoutesCacheKey);
    if (response.expired) {
      log.debug('fetching pilot routes from api');
      data = await fetchPilotRouteData();
      await saveResponseToS3(data, pilotRoutesCacheKey);
    } else if (response.data) {
      log.debug('parsing pilot routes from cache');
      //data is compressed, so decompress before can be put in json form
      const decompressedData = zlib.unzipSync(Buffer.from(response.data, 'base64')).toString('utf-8');
      data = JSON.parse(decompressedData) as FeatureCollection;
    }
  } catch (e) {
    if (!data && response?.data) {
      log.warn('parsing expired pilot routes from cache');
      //data is compressed, so decompress before can be put in json form
      const decompressedData = zlib.unzipSync(Buffer.from(response.data, 'base64')).toString('utf-8');
      data = JSON.parse(decompressedData) as FeatureCollection;
    }
  }

  return data ?? [];
}

// map the routes to fit the db model
function mapPilotRoutes(pilotRoutes: PilotRoute[], features: FeatureCollection) {
  const filteredRoutes: PilotRoute[] = [];

  pilotRoutes.forEach((route) => {
    const foundFeature = features?.features?.find((f) => f.id === route.id);
    if (foundFeature) {
      filteredRoutes.push({
        id: route.id,
        name: foundFeature?.properties?.name,
      } as PilotRoute);
    }
  });
  // if empty array is returned, it means that routes are not found
  // from api or cache
  return filteredRoutes ?? [];
}

function mapTemporaryNotifications(notifications: TemporaryNotification[]) {
  return notifications.map((notification) => {
    return {
      content: notification.content,
      startDate: notification.startDate,
      endDate: notification.endDate,
    };
  });
}
