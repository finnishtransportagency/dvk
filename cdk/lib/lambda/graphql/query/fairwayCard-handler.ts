import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import { FairwayCard, QueryFairwayCardArgs } from '../../../../graphql/generated';
import { log } from '../../logger';
import FairwayCardDBModel from '../../db/fairwayCardDBModel';
import { FairwayService } from '../../api/fairwayService';

const fairwayService = new FairwayService();

export const handler: AppSyncResolverHandler<QueryFairwayCardArgs, FairwayCard | undefined> = async (
  event: AppSyncResolverEvent<QueryFairwayCardArgs>
): Promise<FairwayCard | undefined> => {
  log.info(`fairwayCard(${event.arguments.id})`);
  const dbModel = await FairwayCardDBModel.get(event.arguments.id);
  if (dbModel) {
    const card: FairwayCard = {
      id: dbModel.id,
      name: {
        fi: dbModel.name?.fi || '',
        sv: dbModel.name?.sv || '',
        en: dbModel.name?.en || '',
      },
      fairways: [],
    };
    for (const fairway of dbModel?.fairways || []) {
      card.fairways.push(fairwayService.mapModelsToFairway(undefined, fairway));
    }
    return card;
  }
  return undefined;
};
