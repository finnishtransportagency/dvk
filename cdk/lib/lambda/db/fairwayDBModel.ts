import { GetCommand, ScanCommand } from '@aws-sdk/lib-dynamodb';
import { log } from '../logger';
import { getDynamoDBDocumentClient } from './dynamoClient';

const fairwayTable = process.env.FAIRWAY_TABLE;

class FairwayDBModel {
  id: number;

  name?: string;

  geotiffImages?: string[];

  static async get(id: number): Promise<FairwayDBModel | undefined> {
    const response = await getDynamoDBDocumentClient().send(new GetCommand({ TableName: fairwayTable, Key: { id } }));
    const fairway = response.Item as FairwayDBModel | undefined;
    log.debug('Fairway name: %s', fairway?.name);
    return fairway;
  }

  static async getAll(): Promise<FairwayDBModel[]> {
    const response = await getDynamoDBDocumentClient().send(new ScanCommand({ TableName: fairwayTable }));
    const fairways = response.Items as FairwayDBModel[] | undefined;
    if (fairways) {
      log.debug('%d fairway(s) found', fairways.length);
      return fairways;
    }
    log.debug('No fairways found');
    return [];
  }
}

export default FairwayDBModel;
