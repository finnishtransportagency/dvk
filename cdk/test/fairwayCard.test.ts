import { mockClient } from 'aws-sdk-client-mock';
import { DynamoDBDocumentClient, GetCommand } from '@aws-sdk/lib-dynamodb';
import { handler } from '../lib/lambda/graphql/query/fairwayCard-handler';
import { Status } from '../graphql/generated';
import { mockContext, mockQueryFairwayCardArgsEvent } from './mocks';
import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { sdkStreamMixin } from '@aws-sdk/util-stream';
import { createReadStream } from 'fs';
import FairwayCardDBModel from '../lib/lambda/db/fairwayCardDBModel';
import { pilotPlaceMap } from '../lib/lambda/db/modelMapper';

const ddbMock = mockClient(DynamoDBDocumentClient);
const s3Mock = mockClient(S3Client);

const card: FairwayCardDBModel = {
  id: 'test',
  name: {
    fi: 'Testfi',
    sv: 'Testsv',
    en: 'Testen',
  },
  creator: 'test',
  creationTimestamp: Date.now(),
  modifier: 'test2',
  modificationTimestamp: Date.now(),
  status: Status.Public,
  fairways: [{ id: 1, primary: true, secondary: false }],
  trafficService: {
    pilot: {
      places: [
        {
          id: 681017200,
        },
      ],
    },
  },
};
const card2: FairwayCardDBModel = {
  id: 'test',
  name: {
    fi: 'Testfi2',
    sv: 'Testsv2',
    en: 'Testen2',
  },
  creator: 'test2',
  creationTimestamp: Date.now(),
  modifier: 'test',
  modificationTimestamp: Date.now(),
  status: Status.Draft,
  fairways: [{ id: 1, primary: true, secondary: false }],
  trafficService: {
    pilot: {
      places: [
        {
          id: 3458100305,
        },
      ],
    },
  },
};
const points = {
  features: [
    {
      type: 'Feature',
      id: 'PilotBoardingPlace_P.fid--73ae6a7c_18944e5de2b_b6',
      geometry: { type: 'Point', coordinates: [28.17, 61.5228333, 0] },
      geometry_name: 'GEOM',
      properties: {
        IDENTIFIER: 'FI 0000034581 00305',
        RWF_ID: '1410492',
        REP_ID: '2049602',
        REP_USAGE_ID: '2425332',
        CATPIL: '3',
        COMCHA: null,
        DATEND: null,
        DATSTA: null,
        INFORM: 'www.pilotorder.fi',
        NINFOM: null,
        NOBJNM: null,
        NPLDST: null,
        NTXTDS: null,
        OBJNAM: 'Puumala2',
        PEREND: null,
        PERSTA: null,
        PILDST: null,
        SORIND: 'FI,FI,reprt,NtM 24/618/2006',
        STATUS: null,
        SPATIAL_ID: null,
        SYMBOLIZATION: null,
        ACRONYM: 'PILBOP',
        PVM: '2023-07-10T06:10:39',
      },
    },
  ],
};

jest.mock('../lib/lambda/environment', () => ({
  getEnvironment: () => 'mock',
  isPermanentEnvironment: () => false,
  getFairwayCardTableName: () => 'FairwayCard-mock',
}));

jest.mock('../lib/lambda/api/axios', () => ({
  fetchTraficomApi: () => points,
}));

beforeEach(() => {
  jest.resetAllMocks();
  ddbMock.reset();
  s3Mock.reset();
  pilotPlaceMap.clear();
});

it('should get card by id from the DynamoDB', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: card,
    });
  const stream = sdkStreamMixin(createReadStream('./test/data/pilotplaces.json'));
  const expires = new Date();
  expires.setTime(expires.getTime() + 1 * 60 * 60 * 1000);
  s3Mock.on(GetObjectCommand).resolves({ Body: stream, Expires: expires });
  const response = await handler(mockQueryFairwayCardArgsEvent, mockContext, () => {});
  expect(response).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

it('should get card by id from the DynamoDB when cache expired', async () => {
  ddbMock
    .on(GetCommand, {
      Key: { id: 'test' },
    })
    .resolves({
      Item: card2,
    });
  const stream = sdkStreamMixin(createReadStream('./test/data/pilotplaces.json'));
  const expires = new Date();
  expires.setTime(expires.getTime() - 1 * 60 * 60 * 1000);
  s3Mock.on(GetObjectCommand).resolves({ Body: stream, Expires: expires });
  const response = await handler(mockQueryFairwayCardArgsEvent, mockContext, () => {});
  expect(response).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});
