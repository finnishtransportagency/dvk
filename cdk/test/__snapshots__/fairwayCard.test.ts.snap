// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get card by id and version by id from the DynamoDB 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": 1,
      "secondary": false,
      "secondarySequenceNumber": 1,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "test",
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen",
    "fi": "Testfi",
    "sv": "Testsv",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 2,
      "name": "Reitti 2",
    },
  ],
  "speedLimit": undefined,
  "status": "DRAFT",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [],
            "type": "Point",
          },
          "id": 681017200,
          "name": {
            "en": "",
            "fi": "",
            "sv": "",
          },
          "pilotJourney": undefined,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v2",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;

exports[`should get latest card by id from the DynamoDB 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": 1,
      "secondary": false,
      "secondarySequenceNumber": 1,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "test",
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen",
    "fi": "Testfi",
    "sv": "Testsv",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 2,
      "name": "Reitti 2",
    },
  ],
  "speedLimit": undefined,
  "status": "DRAFT",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [],
            "type": "Point",
          },
          "id": 681017200,
          "name": {
            "en": "",
            "fi": "",
            "sv": "",
          },
          "pilotJourney": undefined,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0_latest",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;

exports[`should get latest card by id from the DynamoDB when cache expired 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": 1,
      "secondary": false,
      "secondarySequenceNumber": 1,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "test",
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen",
    "fi": "Testfi",
    "sv": "Testsv",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 2,
      "name": "Reitti 2",
    },
  ],
  "speedLimit": undefined,
  "status": "DRAFT",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [],
            "type": "Point",
          },
          "id": 681017200,
          "name": {
            "en": "",
            "fi": "",
            "sv": "",
          },
          "pilotJourney": undefined,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0_latest",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;

exports[`should get public card by id from the DynamoDB 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": 1,
      "secondary": false,
      "secondarySequenceNumber": 1,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "test",
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen",
    "fi": "Testfi",
    "sv": "Testsv",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 1,
      "name": "Reitti 1",
    },
  ],
  "speedLimit": undefined,
  "status": "PUBLIC",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [],
            "type": "Point",
          },
          "id": 681017200,
          "name": {
            "en": "",
            "fi": "",
            "sv": "",
          },
          "pilotJourney": undefined,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0_public",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;

exports[`should get public card when admin role missing 1`] = `
{
  "additionalInfo": undefined,
  "anchorage": undefined,
  "attention": undefined,
  "creationTimestamp": Any<Number>,
  "creator": "test",
  "designSpeed": undefined,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "primarySequenceNumber": 1,
      "secondary": false,
      "secondarySequenceNumber": 1,
    },
  ],
  "generalInfo": undefined,
  "group": undefined,
  "harbors": undefined,
  "iceCondition": undefined,
  "id": "test",
  "lineText": undefined,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "test2",
  "n2000HeightSystem": false,
  "name": {
    "en": "Testen",
    "fi": "Testfi",
    "sv": "Testsv",
  },
  "navigationCondition": undefined,
  "pictures": undefined,
  "pilotRoutes": [
    {
      "id": 1,
      "name": "Reitti 1",
    },
  ],
  "speedLimit": undefined,
  "status": "PUBLIC",
  "temporaryNotifications": [],
  "trafficService": {
    "pilot": {
      "email": undefined,
      "extraInfo": undefined,
      "fax": undefined,
      "internet": undefined,
      "phoneNumber": undefined,
      "places": [
        {
          "geometry": {
            "coordinates": [],
            "type": "Point",
          },
          "id": 681017200,
          "name": {
            "en": "",
            "fi": "",
            "sv": "",
          },
          "pilotJourney": undefined,
        },
      ],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0_public",
  "vesselRecommendation": undefined,
  "visibility": undefined,
  "windRecommendation": undefined,
}
`;
